src/components/tl/ucp/bcast/bcast_knomial.c

```diff
@@ -0,0 +1,86 @@
+        while (_dist*_radix < _size) {          \
+            _dist*=_radix;                      \
+        }                                       \
+    }while(0)
+
+ucc_status_t ucc_tl_ucp_bcast_knomial_progress(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_ucp_task_t *task = ucc_derived_of(coll_task, ucc_tl_ucp_task_t);
+    ucc_tl_ucp_team_t *team       = task->team;
+    uint32_t           myrank     = team->rank;
+    uint32_t           team_size  = team->size;
+    uint32_t           root       = (uint32_t)task->args.root;
+    int                radix      = task->bcast_kn.radix;
+    uint32_t           vrank      = (myrank - root + team_size) % team_size;
+    uint32_t           dist       = task->bcast_kn.dist;
+    void              *buffer     = task->args.src.info.buffer;
+    ucc_memory_type_t  mtype      = task->args.src.info.mem_type;;
+    size_t             data_size  = task->args.src.info.count *
+        ucc_dt_size(task->args.src.info.datatype);
+    uint32_t i, vpeer, peer, vroot_at_level, root_at_level, pos;
+    if (UCC_INPROGRESS == ucc_tl_ucp_test(task)) {
+        return task->super.super.status;
+    }
+    while (dist >= 1) {
+        if (vrank % dist == 0) {
+            pos = (vrank/dist) % radix;
+            if (pos == 0) {
+                for (i = radix - 1; i >= 1; i--) {
+                    vpeer = vrank + i * dist;
+                    if (vpeer < team_size) {
+                        peer = (vpeer + root) % team_size;
+                        ucc_tl_ucp_send_nb(buffer, data_size, mtype, peer, team, task);
+                    }
+                }
+            } else {
+                vroot_at_level = vrank - pos * dist;
+                root_at_level  = (vroot_at_level + root) % team_size;
+                ucc_tl_ucp_recv_nb(buffer, data_size, mtype, root_at_level, team, task);
+                ucc_assert(task->send_posted == 0 &&
+                           task->recv_posted == 1);
+            }
+        }
+        dist /= radix;
+        if (UCC_INPROGRESS == ucc_tl_ucp_test(task)) {
+            task->bcast_kn.dist = dist;
+            return task->super.super.status;
+        }
+    }
+
+    ucc_assert(UCC_TL_UCP_TASK_P2P_COMPLETE(task));
+    task->super.super.status = UCC_OK;
+    return UCC_OK;
+}
+
+ucc_status_t ucc_tl_ucp_bcast_knomial_start(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_ucp_task_t *task = ucc_derived_of(coll_task, ucc_tl_ucp_task_t);
+    ucc_tl_ucp_team_t *team = task->team;
+    ucc_status_t       status;
+    task->bcast_kn.radix =
+        ucc_min(UCC_TL_UCP_TEAM_CTX(team)->cfg.bcast_kn_radix, team->size);
+    CALC_DIST(team->size, task->bcast_kn.radix, task->bcast_kn.dist);
+    task->super.super.status = UCC_INPROGRESS;
+    status = ucc_tl_ucp_bcast_knomial_progress(&task->super);
+    if (UCC_INPROGRESS == status) {
+        ucc_progress_enqueue(UCC_TL_UCP_TEAM_CORE_CTX(team)->pq, &task->super);
+    } else if (status < 0) {
```

Comment:
seems we can remove "else if " and just "return status" from this func

Reply:
status can be "IN_PROGRESS" from the progress function. But we report only UCC_OK or ERROR from ucc_collective_start (we discussed we don't have immediate completions, user always has to call test at least once).